File: apps/cargo-tracker/packs/booking/app/domain/booking/domain/temperature_requirement.rb
# frozen_string_literal: true

module Booking
  module Domain
    # 温度管理条件。最低/最高温度・温度単位（REFRIGERATED 時必須）。
    TemperatureRequirement = Data.define(:min_temperature, :max_temperature, :unit) do
      CELSIUS = "CELSIUS"
      FAHRENHEIT = "FAHRENHEIT"
      UNITS = [ CELSIUS, FAHRENHEIT ].freeze

      def initialize(min_temperature:, max_temperature:, unit:)
        raise ArgumentError, "最低温度は必須です" if min_temperature.nil?
        raise ArgumentError, "最高温度は必須です" if max_temperature.nil?
        raise ArgumentError, "温度単位が不正です: #{unit}" unless UNITS.include?(unit)
        if min_temperature > max_temperature
          raise ArgumentError, "最低温度は最高温度以下である必要があります"
        end

        super
      end
    end
  end
end

File: docs/article/source/haskell/apps/cargo-tracker/src/Cargotracker/Booking/Domain/Model/Value/TemperatureRequirement.hs
{- | 温度管理条件 値オブジェクト (US05, IT2)

冷凍・冷蔵貨物 (Cargo.cargoType == Refrigerated) に必須の追加情報。
- minTemperature: 許容下限温度
- maxTemperature: 許容上限温度 (min <= max)
- temperatureUnit: 'C' (摂氏) or 'F' (華氏)
-}
module Cargotracker.Booking.Domain.Model.Value.TemperatureRequirement
  ( TemperatureRequirement (..),
    TemperatureUnit (..),
    mkTemperatureRequirement,
    parseTemperatureUnit,
  ) where

import Data.Text (Text)

import Cargotracker.Shared.Domain.DomainError (DomainError (..))

data TemperatureUnit
  = Celsius
  | Fahrenheit
  deriving stock (Eq, Show)

data TemperatureRequirement = TemperatureRequirement
  { minTemperature :: !Double
  , maxTemperature :: !Double
  , temperatureUnit :: !TemperatureUnit
  }
  deriving stock (Eq, Show)

parseTemperatureUnit :: Text -> Either DomainError TemperatureUnit
parseTemperatureUnit "C" = Right Celsius
parseTemperatureUnit "F" = Right Fahrenheit
parseTemperatureUnit other =
  Left (InvalidBookingId ("invalid temperature unit: " <> other))

mkTemperatureRequirement ::
  Double ->
  Double ->
  TemperatureUnit ->
  Either DomainError TemperatureRequirement
mkTemperatureRequirement minT maxT unit
  | minT > maxT =
      Left (InvalidBookingId "min temperature must be <= max temperature")
  | otherwise =
      Right
        TemperatureRequirement
          { minTemperature = minT
          , maxTemperature = maxT
          , temperatureUnit = unit
          }
